folio-token/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,141 @@
+      var newToken = login(user.username());
+      user = user.withToken(newToken);
+    } else {
+      user = getSystemUser(tenantId);
+    }
+    systemUserCache.put(tenantId, user);
+
+    return user;
+  }
+
+  /**
+   * Authenticate system user and return token value.
+   *
+   * @param user {@link org.folio.spring.model.SystemUser} to log with
+   * @return token value
+   */
+  public UserToken authSystemUser(SystemUser user) {
+    return getToken(() ->
+            authnClient.loginWithExpiry(new UserCredentials(user.username(), systemUserProperties.password())),
+        user.username(), "log in");
+  }
+
+  @Autowired(required = false)
+  public void setSystemUserCache(Cache<String, SystemUser> systemUserCache) {
+    this.systemUserCache = systemUserCache;
+  }
+
+  private SystemUser getSystemUser(String tenantId) {
+    log.info("Attempting to issue token for system user [tenantId={}]", tenantId);
+    var systemUser = SystemUser.builder()
+        .tenantId(tenantId)
+        .username(systemUserProperties.username())
+        .okapiUrl(environment.getOkapiUrl())
+        .build();
+
+    // create context for authentication
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var token = authSystemUser(systemUser);
+      if (token == null) {
+        loginSystemUser(systemUser);
+      }
+      systemUser = systemUser.withToken(token);
+      log.info("Token for system user has been issued [tenantId={}]", tenantId);
+    }
+    // create context for user with token for getting user id
+    try (var fex = new FolioExecutionContextSetter(contextBuilder.forSystemUser(systemUser))) {
+      var userId = prepareUserService.getFolioUser(systemUserProperties.username())
+          .map(UsersClient.User::id).orElse(null);
+      return systemUser.withUserId(userId);
+    }
+  }
+
+  private UserToken login(String username) {
+    return getToken(() ->
+            authnClient.login(new UserCredentials(username, systemUserProperties.password())),
+        username, "relogin");
+  }
+
+  private UserToken loginSystemUser(SystemUser user) {
+    return getToken(() ->
+            authnClient.login(new UserCredentials(user.username(), systemUserProperties.password())),
+        user.username(), "legacy endpoint");
+  }
+
+  private UserToken getToken(Supplier<ResponseEntity<AuthnClient.LoginResponse>> tokenSupplier,
+                             String username, String action) {
+    var response = tokenSupplier.get();
+
+    if (isNull(response.getBody())) {
+      throw new IllegalStateException(String.format(
+          "User [%s] cannot %s because expire times missing for status %s",
+          username, action, response.getStatusCode()));
+    }
+
+    return Optional.ofNullable(response.getHeaders().get(SET_COOKIE))
+        .filter(list -> !CollectionUtils.isEmpty(list))
+        .map(cookieHeaders -> parseUserTokenFromCookies(cookieHeaders, response.getBody()))
```

Comment:
You are calling this method `getToken` for the `authn/login` endpoint. This endpoint doesn't return cookie headers. This method `getToken` as it is implemented will only work for the new `authn/login-with-expiry` endpoint. The old `authn/login` returns the token in an `X-Okapi-Token` header. Again, please look at Adam's code for guidance. As we have discussed, you need two get token methods, one for the old endpoint and one for the new endpoint. You can keep this method, and use if for `login-with-expiry`. Define a new method called `getTokenLegacy` which gets the token from the header.

Reply:
I see you have created the new `getTokenLegacy` method. That's good. You need that. But you still need to change the two public methods in this class to follow the same logic that Adam has in his code. This logic is as follows:

Try to login using `authn/login-with-expiry`. If the token is null then try to login with `authn/login`.

Reply:
changes done.